### PR TITLE
Add ColorEncoding privately with tests

### LINKED ISSUE #4353: Define `ColorEncoding` with tests
This should be similar to `StringEncoding` except that it encodes colors from features instead of strings.

This will be used to implement style coloring in the points, shapes, vectors, tracks, and maybe all layer types. It will also be used to encode text element colors.

As with `StringEncoding`, this can be private initially, and then exposed when used.

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description

This introduces a new type of style encoding to encode colors from features. It does so privately, so that we can get this first chunk of code in and clarify the basic API. This will be used for `TextManager.color`, as well as layer style properties like `Points.style.face_color` in follow-up PRs.

This will be widely used across the napari codebase, so eventual public names of classes and fields here are especially important. I've used names that are partly inspired by altair (e.g. **nominal** and **quantitative** encodins), but feel free to suggest others.

Also note that behavior of passing a string to a ColorEncoding is consistent with StringEncoding - a direct encoding is created. In altair, doing something similar will effectively generate a nominal or quantitative encoding by inspecting the dtype of the feature column. That's harder to do in napari for 2 reasons: 1) we don't have default categorical colormaps (though maybe we should); and 2) because we'd need `features` to be state associated with `TextManager` and the upcoming `Style` class to make validation/coercion play nicely with our pydantic models.

After writing this PR, I'm also becoming more uncertain about whether using things like `StringArray` and `ColorArray` as return types is meaningful as they're somewhat uninteresting subclasses of `np.ndarray`. They're mostly useful as pydantic model field validation types. I guess one way to think about them is that a return value typed with `ColorArray` should probably be untouched when being passed through `ColorArray.validate_type`. Another thought I had (which might be particularly relevant for colors) is if we want to generalize a style array return value to just be anything that implements `__array__` (e.g. to provide support for storing values directly on VRAM and avoid copying between that and RAM), but I may be overthinking things.

## Type of change
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Closes #4353
Provides the basis for #3884.

For more context on how we expect to use `ColorEncoding` in `TextManager`, see https://github.com/andy-sweet/napari/pull/5 and https://github.com/andy-sweet/napari/pull/6

For more context on the overall design changes here, see: https://hackmd.io/Ybil_epnQPiBczPcbGRqCw?view

# How has this been tested?
- [ ] `test_color_encoding.py` covers most basic usage of color encodings
- [ ] All existing tests continue to pass, even though there is a small change to style encoding models

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by andy-sweet** [comment]
@tlambert03 : I added you as a reviewer to take a first pass at this, as our head pydant.

@jni and @kevinyamauchi may also be interested though, so feel free to take a look and make suggestions.

File: napari/layers/utils/style_encoding.py
Comment (jni) [inline napari/layers/utils/style_encoding.py]: I honestly have no idea what this is doing? Probably why you wanted @tlambert03 to review, but possibly a more extnesive comment and class docstring would be helpful for future code archeologists?

File: napari/layers/utils/style_encoding.py
Comment (jni) [inline napari/layers/utils/style_encoding.py:125]: Can you also briefly explain in a comment above how Generic works? Maybe with a link to the Python typing docs? @tlambert03 has sometimes been frustrated when I ask that comments explain "standard" Python features, but my feeling is that there are standard but *uncommon* features that should be explained when they are used. If someone with my Python experience hasn't encountered them before, then that's a pretty good indicator that others will also be confused when reading the code, and a clarifying comment is warranted.

File: napari/layers/utils/_color_encoding.py
Comment (jni) [inline napari/layers/utils/_color_encoding.py]: I wonder whether, rather than having "call" set state on the encoding, we should instead *always* calculate contrast limits on "call" if they are not set? ie out of all functions in this work, "call" should be imho the most stateless? And indeed, rather than having the encoding take in a table and refer to state, could it be stateless and always take in a series or array?

**Comment by jni** [comment]
I've left some minor comments @andy-sweet. To your comments:

- I think nominal and quantitative are good names
- I do think we should add a default categorical colormap — perhaps even make it the labels colormap, since users have requested a method to ensure that categories and labels colors match. And, regardless of the issues with labels rendering performance, the resulting color values have become something of a napari trademark!
- "if we want to generalize a style array return value to just be anything that implements __array__ (e.g. to provide support for storing values directly on VRAM and avoid copying between that and RAM), but I may be overthinking things." You may, but I like it. 😂

File: napari/layers/utils/_color_encoding.py
Comment (andy-sweet) [reply napari/layers/utils/_color_encoding.py]: Good catch and agreed. I brought this work in from an older PR, which had a slightly different design and this was an oversight on my part. I'll address this before we merge here.

File: napari/layers/utils/style_encoding.py
Comment (andy-sweet) [reply napari/layers/utils/style_encoding.py]: Good call. At a minimum, a link to the pydantic docs on the config would probably be a good idea. But I can also expand the comment.

We might actually want to put this config value in `EventedModel` - it forbids extra initialization parameters, whereas by default pydantic ignores them. This is particularly useful for color encodings because the different types have some overlapping field names.

File: napari/layers/utils/style_encoding.py
Comment (andy-sweet) [reply napari/layers/utils/style_encoding.py:125]: Sure. In my description, I'm slightly questioning the specific types these take on `StringEncoding` and `ColorEncoding`. But ultimately I'm trying to get across that calling these types of encodings will return array likes with certain dtypes and shapes.

**Comment by andy-sweet** [comment]
> * I do think we should add a default categorical colormap

I like it, but then it's worth clarifying what the following would do.

```python
features = {'class': ['A', 'B', 'A']}
points = Points(np.random.rand(3, 2), features=features)
points.style.face_color = 'class'
```

With this PR and the anticipated follow up work to use this in the `Points` class, that would create a `DirectColorEncoding(feature='class')`, which would fail when actually applying the features because that column does not contain color-like values.

Currently, the color manager logic is smarter because it looks at the dtype of the column to infer if it should be a continuous or categorical (i.e. quantitative or nominal), though I do not believe it supports looking up colors directly. This is more similar to vega/altair, but the downside is that you need `features` to do that inference. Previously, I think we've disliked that particular kind of magic (i.e. coercing the value depends on other state), but it does have some nice ergonomics too.

File: napari/layers/utils/style_encoding.py
Comment (andy-sweet) [reply napari/layers/utils/style_encoding.py]: Done. @tlambert03 : not sure if this config value should go in `EventedModel.Config`. I prefer it to the default of `'ignore'`, but it could break current usage elsewhere.

File: napari/layers/utils/style_encoding.py
Comment (andy-sweet) [reply napari/layers/utils/style_encoding.py:125]: Done.

File: napari/layers/utils/_color_encoding.py
Comment (andy-sweet) [reply napari/layers/utils/_color_encoding.py]: Done. We will need to handle refreshing the contrasting limits if we want to implement napari's current behavior, but I think that's better done in a PR that actually uses this (e.g. using `ColorEncoding` in `Points`).

**Comment by andy-sweet** [comment]
Another related idea is if we want to support a similar shorthand to altair to denote what type of encoding to use. I.e. when coercing a color encoding from a string, something like `'class:q'` would denote that you want `QuantitativeColorEncoding(feature='class')`.

**Review by tlambert03 — APPROVED** [review]
beautiful as always ❤️.  Nice to start seeing how StyleEncoding gets extended.  Typing and documentation are gorgeous.  I'm fine with all the naming decisions (and, as you say, we can still give them more additional consideration before making public)

thanks @andy-sweet

File: napari/layers/utils/style_encoding.py
Comment (tlambert03) [reply napari/layers/utils/style_encoding.py]: yeah, i'd be ok with it being the default as well (but, wouldn't be surprised if things broke if it got changed)

**Comment by andy-sweet** [comment]
> > * I do think we should add a default categorical colormap
> 
> I like it, but then it's worth clarifying what the following would do.
> 
> ```python
> features = {'class': ['A', 'B', 'A']}
> points = Points(np.random.rand(3, 2), features=features)
> points.style.face_color = 'class'
> ```
> 
> With this PR and the anticipated follow up work to use this in the `Points` class, that would create a `DirectColorEncoding(feature='class')`, which would fail when actually applying the features because that column does not contain color-like values.
> 
> Currently, the color manager logic is smarter because it looks at the dtype of the column to infer if it should be a continuous or categorical (i.e. quantitative or nominal), though I do not believe it supports looking up colors directly. This is more similar to vega/altair, but the downside is that you need `features` to do that inference. Previously, I think we've disliked that particular kind of magic (i.e. coercing the value depends on other state), but it does have some nice ergonomics too.

@jni : is adding a default categorical colormap blocking this PR? If so, I need more feedback on the point above to understand what the desired behavior when being able to distinguish different types of color encodings from feature strings or dicts.

**Review by jni — APPROVED** [review]
@andy-sweet no it's not blocking, let's leave it for a future PR.

Forgive me if this is rehashing old discussions, but I'm wondering whether we should separate encodings, which map arrays of values to styles, and something like "table style encodings", which map tables + feature names to styles. ie, to me, it's a little bit weird/"impure" that the column name (if any) is stored in the encoding, and then the input is a table, as opposed to the input being an array. Potentially, one could have a multilevel thing, *or*, there is a convenience API, where an encoding can take *either* a table+column name *or* an array, *and* the column name can optionally be stored ahead of time.

Anyway, these things can be discussed as we make more of this API public, but that is the thought I had while reading this...

**Comment by andy-sweet** [comment]
> Forgive me if this is rehashing old discussions, but I'm wondering whether we should separate encodings, which map arrays of values to styles, and something like "table style encodings", which map tables + feature names to styles. ie, to me, it's a little bit weird/"impure" that the column name (if any) is stored in the encoding, and then the input is a table, as opposed to the input being an array. Potentially, one could have a multilevel thing, _or_, there is a convenience API, where an encoding can take _either_ a table+column name _or_ an array, _and_ the column name can optionally be stored ahead of time.

I think we've had a similar discussion before, but always feel free to bring up issues like this. I think have reasonable answers below, but maybe I'm in too deep here!

To me a pure function always returns the same values given the same input, which is true for all our instances of `StyleEncoding.__call__` so far and should continue to be. That function also doesn't have side-effects on any of its inputs or anything else, which also seems like a good characteristic.

If you don't store the column name in the encoding, where should it be stored? It seems like we need to reference it later (e.g. when a new point is added) and also want to serialize it for more persistent storage (e.g. with the rest of the layer). Storing it directly as an attribute on a layer seems worse than storing it in the encoding because there's already so much there and it's unclear how we'd encapsulate the state of the encoding as we do now.

Maybe the weirdness here is because the features table is often way more than we need to encode the style values? I see that, but we need it for `FormatStringEncoding` and may want it for other types of encodings in the future (e.g. a numeric encoding that is some combination of many features, or a boolean encoding that checks many features). One other benefit of making that the general signature for `StyleEncoding.__call__` is that the caller (e.g. `Layer` in napari's case) can always call something like `self.style.face_color(self.features)` no matter what type of the encoding it actually is, which significantly reduces the complexity of the code in our layers.

**Comment by sofroniewn** [comment]
> @andy-sweet no it's not blocking, let's leave it for a future PR.

Sounds like this is ready to go in now! Unless hearing anything else, I will merge tomorrow - thanks!!

**Comment by jni** [comment]
Oops, sorry that I missed your update over the weekend, @andy-sweet!

> Maybe the weirdness here is because the features table is often way more than we need to encode the style values?

Bingo! Yeah that's the bit that's weird to me, and that makes me think there is a simpler way lurking somewhere. But,

> I see that, but we need it for FormatStringEncoding and may want it for other types of encodings in the future

That is a very good point, and that's probably when we had the discussion before! 😅 

One thing I can imagine is that someone wants to color their own array of values using the same encoding that we are using for face_color, for example to make a colorbar. One thing we might try is having singledispatch on `__call__` and either apply the encoding directly if we get a NumPy array or pandas Series or other column-like, or use the feature access if we get a table-like.

But, yeah, that's the use case I'm imagining, and that is more natural if there is e.g. a TableEncoding thing that stores the column(s), and that contains an ArrayEncoding that does not, but *does* contain the contrast limits. For example! However, that is something that *could* be supported after this merges, so let's press on!

@sofroniewn maybe leave the merge until after the architecture meeting (4pm PDT)? Then @andy-sweet and I can discuss the above. I suspect we'll merge anyway but good to have a sync first.

**Comment by sofroniewn** [comment]
I think this can go in now - it's private so we can always adjust in follow up PRs - thanks all!!

**Pull request merged**
